driver/src/orderbook/shadow_orderbook.rs

```diff
@@ -0,0 +1,312 @@
+            .user_token_pairs()
+            .chain(shadow.user_token_pairs())
+            .collect::<HashSet<_>>();
+
+        let mut changes = Vec::new();
+        for (user, token_id) in user_token_pairs {
+            let primary_balance = primary.read_balance(token_id, user);
+            let shadow_balance = shadow.read_balance(token_id, user);
+            if primary_balance != shadow_balance {
+                changes.push(BalanceChange {
+                    user,
+                    token: token_id.into(),
+                    primary: primary_balance,
+                    shadow: shadow_balance,
+                })
+            }
+        }
+
+        changes
+    }
+}
+
+/// Represents a change in order data between a primary and shadow orderbook.
+#[derive(Debug, PartialEq)]
+enum OrderChange {
+    /// An order was added, i.e. it exists in the primary but not the shadow
+    /// orderbook.
+    Added(Order),
+    /// An order was removed, i.e. it is missing from the primary orderbook but
+    /// exists in shadow orderbook.
+    Removed(Order),
+    /// An order was modified, i.e. it exists in both orderbooks but with
+    /// different values.
+    Modified {
+        user: Address,
+        id: u16,
+        primary: OrderValues,
+        shadow: OrderValues,
+    },
+}
+
+impl OrderChange {
+    /// Compares a primary and shadow orderbook order vector and return the
+    /// detected changes.
+    fn compare_orders(primary: &[Order], shadow: &[Order]) -> Vec<OrderChange> {
+        let mut shadow_orders = shadow
+            .iter()
+            .map(|order| ((order.account_id, order.id), order))
+            .collect::<HashMap<_, _>>();
+
+        let mut changes = Vec::new();
+        for primary_order in primary {
+            let (user, id) = (primary_order.account_id, primary_order.id);
+            let change = match shadow_orders.remove(&(user, id)) {
+                None => OrderChange::Added(primary_order.clone()),
+                Some(shadow_order) if primary_order != shadow_order => OrderChange::Modified {
+                    user,
+                    id,
+                    primary: OrderValues::from(primary_order),
+                    shadow: OrderValues::from(shadow_order),
+                },
+                _ => continue,
+            };
+            changes.push(change);
+        }
+
+        for (_, remaining_shadow_order) in shadow_orders {
+            changes.push(OrderChange::Removed(remaining_shadow_order.clone()))
+        }
+
+        changes
+    }
+
+    /// Retrieves the user and order ID for the order change.
+    #[cfg(test)]
+    fn user_and_id(&self) -> (Address, u16) {
+        match self {
+            OrderChange::Added(order) | OrderChange::Removed(order) => (order.account_id, order.id),
+            OrderChange::Modified { user, id, .. } => (*user, *id),
+        }
+    }
+}
+
+/// Values that can possibly differ between orders.
+#[derive(Debug, PartialEq)]
+struct OrderValues {
```

Comment:
I'd be fine reusing the Order struct since address and userId will also match, but you are right that this is a little cleaner.

It would be good to also have the validity information on these orders (to fetch race condition with cancellations) but I suppose this would involve more piping work as we currently don't populate the on the orders.

Reply:
I wanted the type to capture that only those values can be different between orders (since user ID and order ID are used to identify them).

Yeah, we currently don’t provide validity information from the readers. I think it’s fairly easy to implement though, so I’ll add it as a separate PR.